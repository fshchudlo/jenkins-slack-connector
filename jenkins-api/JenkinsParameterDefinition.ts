export type JenkinsParameterDefinition = {
    type: "StringParameterDefinition" | "TextParameterDefinition" | "ChoiceParameterDefinition" | "BooleanParameterDefinition";
    name: string;
    description?: string;
    defaultParameterValue?: { value: any };
    definition?: {
        choices?: string[];
        defaultVal?: any;
    };
    choices?: string[];
}
