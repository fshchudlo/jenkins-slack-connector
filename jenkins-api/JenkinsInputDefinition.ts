import {JenkinsParameterDefinition} from "./JenkinsParameterDefinition";

export type JenkinsInputDefinition = {
    id: string;
    message: string;
    proceedText: string;
    proceedUrl: string;
    abortUrl: string;
    inputs: JenkinsParameterDefinition[];
}