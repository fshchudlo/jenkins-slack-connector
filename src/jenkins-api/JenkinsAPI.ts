import axios, {AxiosResponse, Method} from "axios";
import {BasicAuthCredentials} from "./BasicAuthCredentials";
import {JenkinsInputDefinition} from "./JenkinsInputDefinition";
import {JenkinsParameterDefinition} from "./JenkinsParameterDefinition";

const JENKINS_API_POSTFIX = "/api/json";

export class JenkinsAPI {
    private readonly auth: BasicAuthCredentials;
    private readonly jenkinsUrl: string;

    constructor(auth: BasicAuthCredentials, jenkinsUrl: string) {
        this.auth = auth;
        this.jenkinsUrl = jenkinsUrl;
    }

    private async request(url: string, method: Method = "GET", params?: any): Promise<AxiosResponse> {
        return axios.request({
            method,
            url,
            params,
            auth: this.auth,
            headers: {"accept": "application/json", "accept-encoding": "gzip, deflate, br"}
        });
    }

    async startJob(jobName: string, jobParams?: any): Promise<string> {
        const runUrlPart = jobParams ? "buildWithParameters" : "build";
        const runJobUrl = `${this.jenkinsUrl}/${jobName}/${runUrlPart}`;

        const queuedJobUrl = (await this.request(runJobUrl, "POST", jobParams)).headers.location;

        // give Jenkins time to start the job
        await new Promise(resolve => setTimeout(resolve, 300));

        const startedJob = (await this.request(`${queuedJobUrl}${JENKINS_API_POSTFIX}`, "GET")).data;
        if (startedJob.blocked) {
            return Promise.reject(startedJob.why);
        }
        return startedJob.executable.url;
    }

    async getJobParameters(jobPath: string): Promise<JenkinsParameterDefinition[] | null> {
        const url = `${this.jenkinsUrl}/${jobPath}/${JENKINS_API_POSTFIX}`;
        const buildInfo = (await this.request(url, "GET")).data;
        const parametersDefinition = buildInfo.property.find((item: object) => "parameterDefinitions" in item);
        return parametersDefinition?.parameterDefinitions || null;
    }

    async getJobPendingInput(jobUrl: string): Promise<JenkinsInputDefinition> {
        const url = `${jobUrl}/wfapi/nextPendingInputAction`;
        return (await this.request(url, "GET")).data;
    }

    async abortJobPendingInput(abortUrl: string): Promise<void> {
        const fullAbortUrl = `${this.jenkinsUrl}/${abortUrl}`;
        await this.request(fullAbortUrl, "POST");
    }

    async submitJobPendingInput(buildUrl: string, submitUrl: string, formValues: any): Promise<void> {
        const inputDefinition = await this.getJobPendingInput(buildUrl);

        if (!inputDefinition || !submitUrl.includes(`inputSubmit?inputId=${inputDefinition.id}`)) {
            throw new Error("Input was already submitted or timed out");
        }
        this.appendUnselectedCheckboxes(formValues, inputDefinition);

        const convertedValues = Object.entries(formValues).map(([key, value]) => ({name: key, value}));
        await this.request(`${this.jenkinsUrl}${submitUrl}`, "POST", {json: JSON.stringify({parameter: convertedValues})});
    }

    private appendUnselectedCheckboxes(formValues: any, inputDefinition: JenkinsInputDefinition): Record<string, any> {
        inputDefinition.inputs.forEach(parameter => {
            if (parameter.type === "BooleanParameterDefinition" && !(parameter.name in formValues)) {
                formValues[parameter.name] = false;
            }
        });
        return formValues;
    }
}
