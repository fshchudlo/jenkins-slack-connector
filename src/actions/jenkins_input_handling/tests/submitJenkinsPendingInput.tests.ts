import {submitJenkinsPendingInput} from "../../index";
import {JenkinsAPI} from "../../../jenkins-api/JenkinsAPI";

jest.mock("../../../jenkins-api/JenkinsAPI");

describe("submitJenkinsPendingInput", () => {
    let context: any;
    let body: any;
    let ack: jest.Mock;
    let respond: jest.Mock;

    beforeEach(() => {
        context = {jenkinsAPI: new JenkinsAPI(null, null)};
        body = {
            actions: [{value: "http://example.com/submit", block_id: "http://example.com/build"}],
            state: {
                values: {
                    "someStringParam": {
                        "someStringParam": {
                            "type": "plain_text_input",
                            "value": "test"
                        }
                    }
                }
            },
            user_name: "test_user"
        };
        ack = jest.fn();
        respond = jest.fn();
    });

    it("responds with submitted values", async () => {
        jest.spyOn(context.jenkinsAPI, "submitJobPendingInput").mockResolvedValueOnce(undefined);

        await submitJenkinsPendingInput({context, body, ack, respond});

        expect(context.jenkinsAPI.submitJobPendingInput).toHaveBeenCalledWith("http://example.com/build", "http://example.com/submit", {someStringParam: "test"});
        expect(respond).toHaveBeenCalled();
        expect(respond.mock.calls[0][0]).toMatchSnapshot();
    });

    it("responds without submitted values", async () => {
        body.state.values = {}
        jest.spyOn(context.jenkinsAPI, "submitJobPendingInput").mockResolvedValueOnce(undefined);

        await submitJenkinsPendingInput({context, body, ack, respond});

        expect(context.jenkinsAPI.submitJobPendingInput).toHaveBeenCalledWith("http://example.com/build", "http://example.com/submit", {});
        expect(respond).toHaveBeenCalled();
        expect(respond.mock.calls[0][0]).toMatchSnapshot();
    });

    it("responds with an error message if submission fails", async () => {
        const error = new Error("Network error");
        jest.spyOn(context.jenkinsAPI, "submitJobPendingInput").mockRejectedValueOnce(error);

        await submitJenkinsPendingInput({context, body, ack, respond});

        expect(context.jenkinsAPI.submitJobPendingInput).toHaveBeenCalledWith("http://example.com/build", "http://example.com/submit", {someStringParam: "test"});
        expect(respond).toHaveBeenCalled();
        expect(respond.mock.calls[0][0]).toMatchSnapshot();
    });
});