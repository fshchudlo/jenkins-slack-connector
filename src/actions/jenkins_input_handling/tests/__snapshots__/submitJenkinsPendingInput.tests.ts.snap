// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`submitJenkinsPendingInput responds with an error message if submission fails 1`] = `
{
  "blocks": [
    {
      "text": {
        "text": ":octagonal_sign: *Unable to submit input for the <http://example.com/build|build>* 
\`Error: Network error\`",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
}
`;

exports[`submitJenkinsPendingInput responds with submitted values 1`] = `
{
  "blocks": [
    {
      "text": {
        "text": "Input was submitted to the <http://example.com/build|build> by test_user",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "elements": [
        {
          "text": "Submitted values:",
          "type": "plain_text",
        },
      ],
      "type": "context",
    },
    {
      "elements": [
        {
          "text": "	â€¢ someStringParam: test",
          "type": "plain_text",
        },
      ],
      "type": "context",
    },
  ],
}
`;

exports[`submitJenkinsPendingInput responds without submitted values 1`] = `"Input was submitted to the <http://example.com/build|build> by test_user"`;
